File: src/bin/0050-powx-n.cpp
// @leetup=info id=50 lang=cpp slug=powx-n

#include <algorithm>
#include <cassert>
#include <iostream>
#include <map>
#include <set>
#include <tuple>
#include <vector>

// @leetup=code

class Solution {
 public:
  double myPow(double x, int n) {
    if (n == 0) {
      return 1.0;
    }
    if (n > 0) {
      return pow(x, n);
    }
    return 1.0 / pow(x, -static_cast<int64_t>(n));
  }

 private:
  double pow(double x, uint64_t n) {
    double res = 1.0;
    double p = x;
    while (n != 0) {
      if (n & 1) {
        res *= p;
      }
      n >>= 1;
      p *= p;
    }
    return res;
  }
};
// @leetup=code

int main() {
  std::cout << "Hello, world!" << std::endl;
  return 0;
}

File: src/bin/0050-powx-n.py
# @leetup=info id=50 lang=python3 slug=powx-n



# @leetup=code
class Solution:
    def myPow(self, x: float, n: int) -> float:
        def pow(x: float, n: int) -> float:
            if n == 1:
                return x
            if n % 2 == 0:
                return self.myPow(x * x, n // 2)
            return self.myPow(x * x, n // 2) * x

        if n == 0:
            return 1
        if n < 0:
            return 1.0 / pow(x, -n)
        return pow(x, n)


# @leetup=code
